File: src/backend/call-logger/app/services/record_service.py
"""
DynamoDB CRUD for call records.

Table:  {env}-driver-assistant-call-records
PK:     call_id  (String) — LiveKit room name, e.g. driver-call-_+84867347452_abc
GSI:    caller-started_at-index  — query all calls by driver phone number
"""
import json
from decimal import Decimal
from functools import lru_cache

import boto3
from boto3.dynamodb.conditions import Key

from app.core.config import get_settings


@lru_cache
def _table():
    s = get_settings()
    return boto3.resource("dynamodb").Table(s.dynamodb_table)


def _to_decimal(obj: dict) -> dict:
    """DynamoDB doesn't accept float — convert via JSON round-trip."""
    return json.loads(json.dumps(obj), parse_float=Decimal)


def save(record: dict) -> None:
    _table().put_item(Item=_to_decimal(record))


def get(call_id: str) -> dict | None:
    resp = _table().get_item(Key={"call_id": call_id})
    return resp.get("Item")


def list_by_caller(caller: str, limit: int = 20) -> list[dict]:
    """All calls for one driver, newest first."""
    resp = _table().query(
        IndexName="caller-started_at-index",
        KeyConditionExpression=Key("caller").eq(caller),
        ScanIndexForward=False,
        Limit=limit,
    )
    return resp.get("Items", [])


def list_recent(limit: int = 20) -> list[dict]:
    """Scan for recent records — dev/debug only, avoid on large tables."""
    resp = _table().scan(Limit=limit)
    items = resp.get("Items", [])
    return sorted(items, key=lambda x: x.get("started_at", ""), reverse=True)

File: src/backend/call-logger/src/services/recordService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
  QueryCommand,
  ScanCommand,
} from '@aws-sdk/lib-dynamodb';
import { config } from '../config';

const docClient = DynamoDBDocumentClient.from(
  new DynamoDBClient({ region: config.awsRegion })
);

export interface Turn {
  role: string;
  text: string;
  ts: number;
}

export interface CallRecord {
  call_id: string;
  caller: string;
  started_at: string;
  ended_at: string;
  duration_seconds: number;
  turns: Turn[];
  summary: string;
  stt_provider?: string;
  llm_model?: string;
}

export async function save(record: CallRecord): Promise<void> {
  await docClient.send(new PutCommand({
    TableName: config.dynamodbTable,
    Item: record,
  }));
}

export async function get(callId: string): Promise<CallRecord | null> {
  const result = await docClient.send(new GetCommand({
    TableName: config.dynamodbTable,
    Key: { call_id: callId },
  }));
  return (result.Item as CallRecord) ?? null;
}

export async function listByCaller(caller: string, limit = 20): Promise<CallRecord[]> {
  const result = await docClient.send(new QueryCommand({
    TableName: config.dynamodbTable,
    IndexName: 'caller-started_at-index',
    KeyConditionExpression: 'caller = :caller',
    ExpressionAttributeValues: { ':caller': caller },
    ScanIndexForward: false,
    Limit: limit,
  }));
  return (result.Items as CallRecord[]) ?? [];
}

export async function listRecent(limit = 20): Promise<CallRecord[]> {
  const result = await docClient.send(new ScanCommand({
    TableName: config.dynamodbTable,
    Limit: limit,
  }));
  const items = (result.Items as CallRecord[]) ?? [];
  return items.sort((a, b) => b.started_at.localeCompare(a.started_at));
}
